packages/gatsby/src/utils/worker/render-html.ts

```diff
@@ -1,50 +1,315 @@
 import fs from "fs-extra"
-import Promise from "bluebird"
+import Bluebird from "bluebird"
 import { join } from "path"
+import { get, isObject, flatten, uniqBy, concat } from "lodash"
+
 import { getPageHtmlFilePath } from "../../utils/page-html"
+import { IPageDataWithQueryResult } from "../../utils/page-data"
 import { Stage } from "../../commands/types"
 
-export const renderHTML = ({
+let lastRenderHTMLCallTimestamp = 0
+let htmlComponentRenderer
+let webpackStats
+
+const staticQueryResultCache = new Map<string, any>()
+const inFlightStaticQueryPromise = new Map<string, Promise<any>>()
+
+const inlineCssCache = new Map<string, string>()
+const inFlightInlineCssPromise = new Map<string, Promise<string>>()
+
+const resourcesForTemplateCache = new Map<string, IResourcesForTemplate>()
+const inFlightResourcesForTemplate = new Map<
+  string,
+  Promise<IResourcesForTemplate>
+>()
+
+const getStaticQueryPath = (hash: string): string =>
+  join(`page-data`, `sq`, `d`, `${hash}.json`)
+
+const getStaticQueryResult = async (hash: string): any => {
+  const staticQueryPath = getStaticQueryPath(hash)
+  const absoluteStaticQueryPath = join(process.cwd(), `public`, staticQueryPath)
+  const staticQueryRaw = await fs.readFile(absoluteStaticQueryPath)
+
+  return JSON.parse(staticQueryRaw.toString())
+}
+
+async function readPageData(
+  publicDir: string,
+  pagePath: string
+): Promise<IPageDataWithQueryResult> {
+  const fixedPagePath = pagePath === `/` ? `index` : pagePath
+  const filePath = join(publicDir, `page-data`, fixedPagePath, `page-data.json`)
+  const rawPageData = await fs.readFile(filePath, `utf-8`)
+
+  return JSON.parse(rawPageData)
+}
+
+async function readWebpackStats(publicDir: string): any {
+  const filePath = join(publicDir, `webpack.stats.json`)
+  const rawPageData = await fs.readFile(filePath, `utf-8`)
+
+  return JSON.parse(rawPageData)
+}
+
+interface IScriptsAndStyles {
+  scripts: Array<any>
+  styles: Array<any>
+  reversedStyles: Array<any>
+  reversedScripts: Array<any>
+}
+
+async function getScriptsAndStylesForTemplate(
+  componentChunkName
+): Promise<IScriptsAndStyles> {
+  // Create paths to scripts
+  let scriptsAndStyles = flatten(
+    [`app`, componentChunkName].map(s => {
```

Comment:
Nit: I think we can use `["app", "componentChunkName"].flatMap()` now that the minimum Node version is 12.13